pandas/core/indexes/datetimelike.py

```diff
@@ -731,7 +731,9 @@ def astype(self, dtype, copy=True):
             # Ensure that self.astype(self.dtype) is self
             return self
 
-        new_values = self._data.astype(dtype, copy=copy)
+        # use _eadata since _data links to _index_data which can be
+        #  overwritten in groupby-reduction
+        new_values = self._eadata.astype(dtype, copy=copy)
```

Comment:
Can you explain why this is needed in `astype`, while the comment is about groupby?

And it's not possible to prevent that _data gets written in the first place? (or is it done in a try/except that only fails later .. ). Adding yet another data like attribute doesn't seem ideal ..

Reply:
> And it's not possible to prevent that _data gets written in the first place?

Take a look at how we patch `_index_data` in libreduction.  I spent some time trying to find a way around that and didn't find any.

Reply:
Could we make the `_index_data` in a settable property that on setting also updates the underlying data of the EA? (to keep both synced) 

(didn't look in detail, so this might be nonsense)

Reply:
`_index_data` is set using `object.__setattr__`, so I don't think that would work.

Reply:
> Do you have an example test that hits this (a failing test if you don't do the above change) ?

If you make the groupby change to stop catching AttributeError but don't change the index code, there is a test (dont know which off the top of my head) that raises an AttributeError in `DatetimeIndex.map` because `index._data` is an ndarray instead of a DTA.